src/crystal/system/unix/process.cr

```diff
@@ -0,0 +1,252 @@
+require "c/signal"
+require "c/stdlib"
+require "c/sys/resource"
+require "c/unistd"
+
+class Process
+  protected def open_system
+    exists_system?
+  end
+
+  protected def wait_system
+    Crystal::SignalChildHandler.wait(@waitpid, @pid)
+  end
+
+  protected def self.exit_system(status = 0) : NoReturn
+    LibC.exit(status)
+  end
+
+  protected def self.pid_system : Int64
+    LibC.getpid.to_i64
+  end
+
+  protected def self.ppid_system : Int64
+    LibC.getppid.to_i64
+  end
+
+  protected def exists_system?
+    Process.exists_system?(@pid)
+  end
+
+  protected def self.exists_system?(pid : Int64)
+    ret = LibC.kill(pid, 0)
+    return true if ret == 0
+    return false if Errno.value == Errno::ESRCH
+    raise Errno.new("kill")
+  end
+
+  protected def pgid_system : Int64
+    Process.pgid_system(@pid)
+  end
+
+  protected def self.pgid_system(pid : Int64) : Int64
+    ret = LibC.getpgid(pid)
+    raise Errno.new("getpgid") if ret < 0
+    ret.to_i64
+  end
+
+  protected def terminate_system
+    signal Signal::TERM
+  end
+
+  protected def kill_system
+    signal Signal::KILL
+  end
+
+  protected def signal_system(signal : Signal)
+    ret = LibC.kill(@pid, signal.value)
+    raise Errno.new("kill") if ret < 0
+  end
+
+  protected def create_and_exec(command : String, args : (Array | Tuple)?, env : Env?, clear_env : Bool, fork_input : IO::FileDescriptor, fork_output : IO::FileDescriptor, fork_error : IO::FileDescriptor, chdir : String?, reader_pipe : IO::FileDescriptor, writer_pipe : IO::FileDescriptor)
+    if pid = Process.fork_internal(will_exec: true)
+      pid
+    else
+      begin
+        reader_pipe.close
+        writer_pipe.close_on_exec = true
+        Process.exec_internal(command, args, env, clear_env, fork_input, fork_output, fork_error, chdir)
+      rescue ex : Errno
+        writer_pipe.write_bytes(ex.errno)
+        writer_pipe.write_bytes(ex.message.try(&.bytesize) || 0)
+        writer_pipe << ex.message
+        writer_pipe.close
+      rescue ex
+        ex.inspect_with_backtrace STDERR
+        STDERR.flush
+      ensure
+        LibC._exit 127
+      end
+    end
+  end
+
+  protected def self.fork_system : Process
+    {% raise("Process fork is unsupported with multithread mode") if flag?(:preview_mt) %}
+    if pid = fork_internal(will_exec: false)
+      new pid
+    else
+      begin
+        yield
+        LibC._exit 0
+      rescue ex
+        ex.inspect_with_backtrace STDERR
+        STDERR.flush
+        LibC._exit 1
+      ensure
+        LibC._exit 254 # not reached
+      end
+    end
+  end
+
+  protected def self.fork_system : Process?
+    {% raise("Process fork is unsupported with multithread mode") if flag?(:preview_mt) %}
+    if pid = fork_internal(will_exec: false)
+      new pid
+    else
+      nil
```

Comment:
`else` branch is redundant (that's already the implicit behavior)

Reply:
It's just moving the existing, but yeah, it would clean up :)